frontend/packages/dev-console/src/components/pipelineruns/logs/Logs.tsx

```diff
@@ -0,0 +1,137 @@
+import * as React from 'react';
+import { Base64 } from 'js-base64';
+import { Alert } from '@patternfly/react-core';
+import { resourceURL, modelFor, PodKind, ContainerSpec } from '@console/internal/module/k8s';
+import { WSFactory } from '@console/internal/module/ws-factory';
+import { LOG_SOURCE_TERMINATED } from '@console/internal/components/utils';
+import { coFetchText } from '@console/internal/co-fetch';
+import './Logs.scss';
+
+type LogsProps = {
+  resource: PodKind;
+  resourceStatus: string;
+  container: ContainerSpec;
+  render: boolean;
+  autoScroll?: boolean;
+  onComplete: (containerName: string) => void;
+};
+
+const Logs: React.FC<LogsProps> = ({
+  resource,
+  resourceStatus,
+  container,
+  onComplete,
+  render,
+  autoScroll = true,
+}) => {
+  const { name } = container;
+  const { kind, metadata = {} } = resource;
+  const { name: resName, namespace: resNamespace } = metadata;
+  const contentRef = React.useRef<HTMLDivElement>(null);
+  const [error, setError] = React.useState<boolean>(false);
+  const fetching = React.useRef<boolean>(false);
+  const resourceStatusRef = React.useRef<string>(resourceStatus);
+  const onCompleteRef = React.useRef<(name) => void>();
+  onCompleteRef.current = onComplete;
+  const appendMessage = React.useRef<(blockContent) => void>();
+  appendMessage.current = React.useCallback(
+    (blockContent) => {
+      const contentLines = blockContent.split('\n').filter((line) => !!line);
+      if (contentRef.current && contentLines.length >= 0) {
+        const elements = contentLines.map((content) => {
+          const customElement = document.createElement('div');
+          customElement.textContent = content;
+          return customElement;
+        });
+        elements.forEach((element) => {
+          contentRef.current.append(element);
+        });
+        const lastElement = elements[elements.length - 1];
+        if (render && lastElement && autoScroll) {
+          lastElement.scrollIntoView({ behavior: 'smooth', block: 'end' });
+        }
+      }
+    },
+    [autoScroll, render],
+  );
+
+  if (resourceStatusRef.current !== resourceStatus) {
+    resourceStatusRef.current = resourceStatus;
+  }
+
+  React.useEffect(() => {
+    let loaded: boolean = false;
+    let ws: WSFactory;
+    const urlOpts = {
+      ns: resNamespace,
+      name: resName,
+      path: 'log',
+      queryParams: {
+        container: name,
+        follow: 'true',
+      },
+    };
+    if (!fetching.current) {
```

Comment:
Why do you need to check fetching?
Don't you want to restart the process if any of the dependencies of this hook change?
You don't clean up the `appendMessage` so maybe you don't? I'd expect that if a dependency changes, it clear the logs and starts again.

Reply:
it's there to stop opening multiple WebSockets. once a request is made we don't want to open the WebSocket again. If this is not there it results in adding duplicate logs for a step

Reply:
But then the dependencies should be controlling this. If any of the dependencies change, it makes sense to need to request new data because it could be looking at a different endpoint.

Reply:
I'm not sure why any of these `kind, name, resName, resNamespace` would change. I noticed this when appendmessage and other callbacks were part of dependencies. I will check this again if this is needed or not.

Reply:
You should always be thinking how to write your components in isolation based only on the incoming props. However those props change, you should be reacting accordingly in your component.
Always assume that every prop can change. There are few instances where prop changes can be ignored. For example the default value of an uncontrolled component.

Reply:
fixed